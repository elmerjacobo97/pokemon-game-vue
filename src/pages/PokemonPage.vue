<template>
    <h2 v-if="!pokemon">Espere por favor...</h2>
    <!-- <div v-if="pokemon"> -->
    <div v-else>
        <h1>¿Quién es este Pokemon?</h1>

        <PokemonPicture
            :pokemonId="pokemon.id"
            :showPokemon="showPokemon"
        />

        <!-- Escuchar ese evento @selection="checkAnswer" -->
        <PokemonOptions
            :pokemons="pokemonArr"
            @selection="checkAnswer"
        />

        <!-- template es similar a un fragment -->
        <template v-if="showAnswer">
            <h2 class="fade-in">{{ message }}</h2>
            <button
                class="btn-new-game"
                type="button"
                @click="newGame"
            >Nuevo Juego</button>
        </template>
    </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
    name: 'PokemonPage',
    components: {
        PokemonOptions,
        PokemonPicture,
    },
    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: '',
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions();

            // Generar un número aleatorio entre 0 y 3
            const rndInt = Math.floor(Math.random() * 4);
            this.pokemon = this.pokemonArr[rndInt];
        },
        checkAnswer(selectedId) {
            this.showPokemon = true;
            this.showAnswer = true;

            if (selectedId === this.pokemon.id) {
                this.message = `Correcto ${this.pokemon.name}`
            } else {
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame() {
            this.pokemonArr = []
            this.mixPokemonArray()
            this.showPokemon = false
            this.showAnswer = false
            this.pokemon = null
        }
    },
    mounted() {
        this.mixPokemonArray();
    }
}
</script>
