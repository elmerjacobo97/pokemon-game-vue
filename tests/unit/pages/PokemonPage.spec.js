import {shallowMount} from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage.vue";
import {pokemons} from "../../mocks/pokemons.mock";

describe('PokemonPage component', function () {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    })

    test('debe de hacer match con el snapshot', function() {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('debe de llamar el mixPokemonArray al montar', function() {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        const wrapper = shallowMount(PokemonPage);

        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('debe de hacer match con el snapshot cuando cargan los pokemons', function() {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', function() {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        });

        const picture = wrapper.find('pokemon-picture-stub');
        const options = wrapper.find('pokemon-options-stub');

        expect(picture.exists()).toBeTruthy();
        expect(options.exists()).toBeTruthy();

        expect(picture.attributes('pokemonid')).toBe('1');
        expect(options.attributes('pokemons')).toBeTruthy();
    });

    test('pruebas con checkAnswer', async function() {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                }
            }
        });

        await wrapper.vm.checkAnswer(1);

        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBe(true);
        expect(wrapper.find('h2').text()).toBe(`Correcto ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(10);
        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)
    });
});
