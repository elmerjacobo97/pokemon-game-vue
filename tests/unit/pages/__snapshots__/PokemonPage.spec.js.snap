// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<!-- <div v-if="pokemon"> -->
<div>
  <h1>¿Quién es este Pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- Escuchar ese evento @selection="checkAnswer" -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub><!-- template es similar a un fragment -->
  <!--v-if-->
</div>
`;
